invenio_rdm_records/secret_links/models.py

```diff
@@ -0,0 +1,172 @@
+    In case that the database is leaked, all secret links should be revoked.
+    """
+
+    created = db.Column(
+        db.DateTime, nullable=False, default=datetime.utcnow, index=True
+    )
+    """Creation timestamp."""
+
+    expires_at = db.Column(db.DateTime, nullable=True)
+    """Expiration date."""
+
+    permission_level = db.Column(db.String, nullable=False, default="")
+    """Permission level of the link."""
+
+    def allows(self, action):
+        """Check if the secret link covers the specified permission."""
+        # TODO permission hierarchy
+        return self.permission_level == action
+
+    def revoke(self, commit=False):
+        """Revoke (i.e. delete) this secret link."""
+        db.session.delete(self)
+        if commit:
+            db.session.commit()
+
+        link_revoked.send(self)
+
+    @property
+    def need(self):
+        """Get the needs provided by the link."""
+        return LinkNeed(str(self.id))
+
+    @property
+    def extra_data(self):
+        """Load token data stored in token (ignores expiry date of tokens)."""
+        if self.token:
+            return self.load_token(self.token, force=True)["data"]
+
+        return None
+
+    @property
+    def is_expired(self):
+        """Determine if link is expired."""
+        if self.expires_at:
+            return datetime.utcnow() > self.expires_at
+
+        return False
+
+    @classmethod
+    def get_by_token(cls, token):
+        """Get the SecretLink referenced by the specified token."""
+        data = cls.load_token(token)
+        if data:
+            link = cls.query.get(data["id"])
+            return link
+        else:
+            return None
+
+    @classmethod
+    def create(
+        cls,
+        permission_level,
+        extra_data=dict(),
+        expires_at=None,
+        commit=False,
+    ):
+        """Create a new secret link."""
+        is_date = isinstance(expires_at, date)
+        is_datetime = isinstance(expires_at, datetime)
+        if is_date and not is_datetime:
+            expires_at = datetime.combine(expires_at, datetime.min.time())
+
+        with db.session.begin_nested():
+            obj = cls(
+                permission_level=permission_level,
+                expires_at=expires_at,
+                token="",
+            )
+            db.session.add(obj)
+
+        with db.session.begin_nested():
+            # Create a token pointing back to the SecretLink via its ID
+            #
+            # NOTE: the nested transactions are required in order for the
+            #       SecretLink to get an auto-assigned UUID, which can be
+            #       referenced in the token
+            if expires_at is None:
+                serializer = SecretLinkSerializer()
+            else:
+                serializer = TimedSecretLinkSerializer(expires_at=expires_at)
+
+            obj.token = serializer.create_token(obj.id, extra_data)
+
+        if commit:
+            db.session.commit()
```

Comment:
Question: Is the commit needed here? Usually, I try to avoid having a commit inside the data layer as it can have unwanted side-effects

Reply:
no, it was purely for convenience -- now removed, along with the optional commit in `revoke`